### PR TITLE
Add ConfirmDirectMessage Dialog

### PR DESCRIPTION
**Pull request by 2580ayush2580**
Fix [https://github.com/vector-im/element-web/issues/16300](https://github.com/vector-im/element-web/issues/16300)

Screenshot of dialog:-

Signed-off-by: Ayush Kumar 2580ayush2580@gmail.com

### CODE REVIEW

File: src/components/views/dialogs/ConfirmDirectMessageDialog.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        return (
+            <BaseDialog
+            title={this.props.title || _t("Confirm Message")}
+            className="mx_ConfirmDirectMessageDialog"
+            contentId="mx_ConfirmDirectMessageDialog"
+            onFinished={this.props.onFinished}
+            hasCancel={true}
+        >
+                <p>
+                    {_t("Are you sure you wish to start a direct message with the ") + `${this.props.name}` + " ?"}
```
Comment (SimonBrandner) [inline src/components/views/dialogs/ConfirmDirectMessageDialog.js:37]: I don't think there should be an article before a username. So "Are you sure you wish to start a direct message with" should be enough. Also, there probably shouldn't be the space before the question mark: "?"

File: src/components/views/dialogs/ConfirmDirectMessageDialog.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+/*
+ * A dialog for confirming a Direct Message.
+ */
+
+export default class ConfirmDirectMessageDialog extends React.Component {
+    render() {
+        return (
+            <BaseDialog
+            title={this.props.title || _t("Confirm Message")}
```
Comment (SimonBrandner) [inline src/components/views/dialogs/ConfirmDirectMessageDialog.js:30]: Maybe a different string would be better. "Confirm Direct Messages"?

File: src/components/views/right_panel/UserInfo.tsx [hunk 3]
```diff
@@ -132,8 +135,16 @@ async function openDMForUser(matrixClient: MatrixClient, userId: string) {
             createRoomOptions.encryption = true;
         }
     }
+    createRoom(createRoomOptions)
+    return modal.close();
```
Comment (SimonBrandner) [inline src/components/views/right_panel/UserInfo.tsx:139]: I am not really sure, maybe returning `createRoom(createRoomOptions)` would be better.

**Review by SimonBrandner — CHANGES_REQUESTED** [review]
Otherwise LGTM to me! Thank you very much! I am pretty sure this will save my life at some point :joy: 

Also, do you plan to implement the invite reason in another PR or should I try to do it?

File: src/components/views/right_panel/UserInfo.tsx (hunk 3 above)
Comment (SimonBrandner) [reply src/components/views/right_panel/UserInfo.tsx:139]: It's up to you or a member of the dev team

**Comment by 2580ayush2580** [comment]
> @2580ayush2580 Also, do you plan to implement the invite reason in another PR or should I try to do it?

I don't know how it's working and currently, I am learning how it works. If you want to work  on this you may.

**Review by SimonBrandner — APPROVED** [review]

**Comment by 2580ayush2580** [comment]
@t3chguy @jryans Please review my PR

File: src/components/views/dialogs/ConfirmDirectMessageDialog.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        return (
+            <BaseDialog
+            title={this.props.title || _t("Confirm Direct Messages")}
+            className="mx_ConfirmDirectMessageDialog"
+            contentId="mx_ConfirmDirectMessageDialog"
+            onFinished={this.props.onFinished}
+            hasCancel={true}
+        >
+                <p>
+                    {_t("Are you sure you wish to start a direct message with ") + `${this.props.name}` + "?"}
```
Comment (jryans) [inline src/components/views/dialogs/ConfirmDirectMessageDialog.js:37]: Please use a single string for the entire sentence and string variable for the name:

```suggestion
                    {_t("Are you sure you wish to start a direct message with %(name)s?", { name: this.props.name })}
```

This way, localisers for other languages can reorder the entire sentence in any way needed.

**Review by jryans — CHANGES_REQUESTED** [review]
Generally okay technically, but we'll need a Product review of this change.

**Comment by jryans** [comment]
@2580ayush2580 Could you add a screenshot of the dialog in the top comment of the PR for easier review?

**Comment by nadonomy** [comment]
@2580ayush2580 Thanks for this! I've been pouring some thought into this PR for the last couple of days and unfortunately can't merge this. As always though, you're free to maintain & run these changes in your own fork of Element.

While I empathise with https://github.com/vector-im/element-web/issues/16300 this PR puts unnecessary friction in creating DMs for everyone. The impact of this change is too great— it means new users need to confirm they'd like to direct message someone immediately after clicking on a 'Direct Message' interaction on a user profile.

The best solution (also alluded to in https://github.com/vector-im/element-web/issues/16300) would be to improve the styling of the user profile to avoid accidental clicks for all users.

Relatedly, if you do continue to use this in your own fork:

1. I'd recommend to test to ensure this isn't breaking the current implementation of opening an existing DM if it exists.
2. I'd recommend to update copy to be more terse all round— perhaps simplifying the title to 'New Direct Message' and the body copy to 'Start a new Direct Message with user'
3. The call to action button looks inconsistent with others in the app, worth double checking if there's any extraneous margin or padding.

**Comment by SimonBrandner** [comment]
@nadonomy I understand the reasoning here and I agree that improving the styling is a better solution. I am just wondering what do you think about making this dialog opt-in? Something like "Require confirmation before starting a DM" could be added to the settings

**Comment by jryans** [comment]
Given Nad's feedback above, let's close this, as a different approach is needed.

**Pull request closed without merging**